### ISSUE TITLE
Setting framerate to normal causes various issues

### ISSUE THREAD

**Issue by Shoegzer** [libretro/flycast#152]
Setting the "quick menu" framerate option to "normal" causes various issues.  The following is likely to be an incomplete list, there are likely other issues that further testing may help uncover:

* Freezing when pressing pause in House of the Dead 2 - see https://github.com/libretro/reicast-emulator/issues/102
* Black screen vs. "empire interactive" screen in Pro Pinball Trilogy - see https://github.com/libretro/reicast-emulator/issues/116
* Random input drops at main menu / freezing after pressing start after completing first level in Sturmwind - see https://github.com/libretro/reicast-emulator/issues/61
* Freezing when pressing pause in Dynamite Cop - see https://github.com/libretro/reicast-emulator/issues/135

I'm not entirely sure what the menu option does as there is [no documentation yet](https://docs.libretro.com/library/reicast/), but something's amiss.

Tested on linux x64, kernel 4.15 against commit 81787fcaccc50bb3d40ca256c87d7a7d427286d6

**Comment by Awakened0** [comment]
Setting framerate to normal in Bangai-O helped remove most of the background scroll stuttering I was experiencing. I've only tested that game on Win10-NVidia, and didn't notice any issues compared to the default fullspeed.

**Comment by inactive123** [comment]
Is this issue related too?

https://github.com/libretro/reicast-emulator/issues/135

**Comment by Shoegzer** [comment]
@Awakened0 That's odd, I don't observe any background scroll stuttering and I just tested it with fullspeed.  What commit revision are you testing against?  If not latest, can you try that and report results here?  Also, can you indicate whether you are using OIT or non-OIT (using OIT here)?  Thanks.

**Comment by Awakened0** [comment]
Just grabbed the latest Win64 OIT build off the buildbot (c3712fa72659b2cd6bb5a5fc7af6f9d4641643b9). I normally play at 120hz with swap interval 2, but set my monitor to 60hz with swap 1 for this test just to rule out an issue with that.

I load up level 1 and clear out any enemies in the immediate area. Then I just move left and right near the ceiling and watch how how it scrolls. It's usually smooth for a few seconds, but very quickly starts to regularly stutter. Then if I go into core options, switch framerate to normal and resume gameplay, the stuttering is far lessened. There's still a small hitch every ten seconds or so. But at fullspeed the stutters seem to be more noticeable (maybe last a ms longer or something) and happen probably twice as often. 

I'm not getting sound crackles during that test, so I don't think it's performance related (2500k at 4.4GHz paired with a GTX 970, Win10).

**Comment by Shoegzer** [comment]
Interesting @Awakened0.  I just played through all of level 1 to the boss with framerate at "fullspeed" and didn't have a single stutter.  Some suggestions:

* Update your Nvidia drivers to the latest version, or try a different version 
* Turn on RA logging, launch from console and see if you notice anything different between a session with fullscreen vs. normal
* Load up linux on that hardware and test against that

Hoping others can back up your observation - I'm not (yet) convinced it's a problem with fullspeed but if it is, so far it seems tied to Windows builds.

**Comment by barbudreadmon** [comment]
> 2500k at 4.4GHz 

@Awakened0 i wonder if you shouldn't try without overclocking, overclocking has always been known as a vector of unexpected issues.

**Comment by flyinghead** [comment]
From what I understand looking at the code, the framerate option exists because of the differences between how upstream reicast works, and how a libretro core should work.

In upstream reicast, the emulator thread runs an infinite loop that only ends when the emulator is stopped. Instead, a libretro core should provide an entry point that runs the emulator long enough to generate a single frame and return control to libretro, which will display the frame, poll inputs, etc. Then the entry point is called again for the next frame and so on.

The framerate option defines when the emulator will be paused and control returned to libretro:
- In normal mode, the emulator is paused when a frame has been rendered
- In fullspeed mode, the emulator is paused when a V-blank interrupt is generated

The problem with normal mode is when the game doesn't render any frame. This can happen for several seconds when the game loads for example. During this time, libretro will be unresponsive as the inputs are not polled and the screen not refreshed.

Certain games also stop rendering any frame when you press pause, as seen in issues #61, #102 and #135. In that case, this is a dead lock since the game is waiting for a button press, which will never come since libretro is waiting for the emulator to finish before polling inputs.

For these reasons, fullspeed mode is probably better than normal mode in most if not all cases.

I am just suspicious about one line of code, in fullspeed mode, that changes the synchronization between emulator thread and rendering thread, and that potentially can introduce timing issues. @twinaphex can you sched some light on this?
`rend_end_render()` is called to wait for the renderer to finish rendering. In fullspeed mode it is not called until the V-blank interrupt. Any reason for this? I think it could be called in both places (`rend_end_sch` and `os_DoEvents / rend_vblank`, since it does nothing if a rendering is not in progress.

**Comment by Awakened0** [comment]
> @Awakened0 i wonder if you shouldn't try without overclocking, overclocking has always been known as a vector of unexpected issues.

I wouldn't think that would be the issue, since it doesn't happen with other cores. Even heavy ones like beetle-Saturn or higan. Though those are software rendered.

**Comment by Awakened0** [comment]
Interestingly, if I use the 240p Test Suite's scrolling test, I get perfectly smooth scrolling with either normal or fullspeed settings. This is with "sync to exact content refresh" on with a G-Sync display. But Bangai-O still seems to lurch more on the fullspeed setting vs. normal. It's been so long since I played it on real hardware; I wonder if that's just how it was...

**Comment by Awakened0** [comment]
Hmm, if I switch the video mode in the test suite to 480i (scaled assets), then I get the same kind of stutter I do in Bangai-O. Normal is better than fullspeed, but still not as perfect as 240p mode. I'm testing at the default 640x480 internal res.

**Comment by Shoegzer** [comment]
@Awakened0 Again I get smooth scrolling and no lurching with fullspeed after playing for quite a while.  Just tested latest commit under Linux 4.18 kernel and NVIDIA GTX 1070M and 390.116 drivers.  It has to be something on your system.

@flyinghead Would you consider removing the normal option as a UI setting given the accuracy-related problems we've both observed that it causes?  Those are long-standing issues and could be closed and future ones prevented with the option removed.   Outside of the synchronization issue with `rend_end_render()` you mentioned above, which seems to be minor, I think the fullspeed option could fully replace it.  What do you think?

**Comment by Awakened0** [comment]
>  It has to be something on your system.

Maybe it's a Windows thing? I'm on Win10 with the recently released 430.64 drivers. IIRC, @Tatsuya79 has also mentioned frame pacing issues with Reicast in Win7.

**Comment by Shoegzer** [comment]
Certainly possible but unlikely.  It would have been more widely reported than one or two users.  Try testing on a different system entirely.

**Comment by flyinghead** [comment]
I tested with 240pSuite and I get pretty much the same lurching in the Scrolling Test with "normal" and "fullspeed" (linux).
The only way (on my system) to have a perfectly fluid scrolling is to use "Threaded Rendering".

However I'm not opposed to removing the "normal" option and setting the default to fullspeed.

**Comment by Shoegzer** [comment]
@flyinghead interesting, I don't observe any lurching and Threaded Rendering option is off here.  I can try turning it on to see if our definitions of "lurching" are really nuanced. ;)

Thanks for considering removing the normal option.  The fewer options the better, especially in this case.  I'll close this issue and others related to normal mode as linked, if/when that's done.

**Comment by flyinghead** [comment]
`The fewer options the better`

I'm 200% with you here

**Comment by Awakened0** [comment]
Threaded Rendering runs at double speed or so on my system for some reason, and it'll stutter periodically even at 240p.

I was hoping 240p mode running smooth while 480i isn't would be a helpful clue.

**Comment by flyinghead** [comment]
Make sure to also enable "Synchronous Rendering" along with "Threaded Rendering"

**Comment by Awakened0** [comment]
With Synchronous and Threaded on, it still runs at double speed at 240p mode. 480i runs at normal speed. Neither are perfectly smooth, though 240p is still smoother with those settings enabled.

**Comment by flyinghead** [comment]
I made some videos on my system. While the Threaded Rendering is not perfect, it's still the best option IMO. Like I said virtually no lurching in this mode when not recording video.

Threaded rendering:
https://www.youtube.com/watch?v=V6lyYXvIxfQ

normal rate:
https://www.youtube.com/watch?v=V6lyYXvIxfQ

fullspeed rate:
https://www.youtube.com/watch?v=6xlJKTmTTmc

**Comment by Tatsuya79** [comment]
```
However I'm not opposed to removing the "normal" option and setting the default to fullspeed.
```
Please don't.
The only setup that is almost stutter free for me is normal, threaded off.
(win7 x64 nvidia 770)

**Comment by Shoegzer** [comment]
@Tatsuya79 Leaving normal mode as-is only invites headaches for devs as its inaccuracies yield problems like the ones you see in this issue summary and in many others, some of which are linked from here.  You two seem to be the only ones with this display issue, and even in the unlikely event there are no current ways around it, there will probably be future efforts to get fullscreen mode working even better anyway.

In short, the goal here should be to remove rather than maintain/support old options that provide marginal benefit to some and are known to cause many issues for all.  This permits focus where it belongs.

**Comment by Awakened0** [comment]
> I made some videos on my system. While the Threaded Rendering is not perfect, it's still the best option IMO. Like I said virtually no lurching in this mode when not recording video.

Can you try these in 480i mode (video mode at the bottom of the test suite, scaled assets so it still uses the full screen)? I'm curious if it's worse than 240p on your system like it is on mine.

It's almost like 480i mode has a different frame limiter than 240p, but that's probably not it. I get 0 lurches in 240p no matter what framerate mode I use with threaded and synchronous off.

**Comment by Awakened0** [comment]
In 480i mode, bringing up the RetroArch menu and closing it will cause the framerate to smooth out for a few seconds before it starts to chunk up again.

Something I just thought of, maybe being able to reduce the audio buffer to 512 would help? The mGBA core had a frame pacing issue where doing that fixed it. And if I turn off Audio Synchronization in RetroArch, the stuttering in Reicast at 480i with fullspeed framerate goes away. The same thing happened with mGBA when the buffer was at 1024.

**Comment by Tatsuya79** [comment]
~~It's been a while I haven't tested the various options, and I can say it's smooth now with threaded and fullspeed... so whatever.~~ :smiley: 
I thought Demolish Fist could run at 60fps vs its default 30fps on "fullspeed" before but it's not the case now, always 30fps.

~~I'm noticing video driver on "glcore" is displaying vertical games upside down.~~ fixed in RA

update: tested again with Giga Wing 2 Naomi and only framerate on "normal" + threaded to off gives me no frame skipping.
Threaded to on generates some stop states where a frame is twinned then 1 get skipped.

**Comment by Awakened0** [comment]
I tried adding a 512 audio buffer option just by modifying one line in libretro.cpp:
`"Audio buffer size; 512|1024|2048",`
But it didn't help with the stuttering when I selected it in core options. Is that all that I would need to change for a 512 audio buffer to work?

**Comment by flyinghead** [comment]
Actually I just found out that the "Audio buffer size" setting is not used anywhere in the code! I guess it was used in the past but the audio implementation was redone.

There's a 512-sample ring buffer used instead and its size is not configurable.

**Comment by Awakened0** [comment]
Huh, maybe it's not related to the audio buffer then. I tested turning off audio sync with Bangai-O and there were still lurches in that, unlike the smooth results I got in the test suite. With threaded and synchronous on, I'm not getting double speed in 240p mode for whatever reason today at least.

I tried fullspeed, threaded and synchronous all enabled on my 60hz TV and got much smoother results there. Bangai-O and the test suite just get occasional lurches. Those settings just aren't working as well on the monitor with G-Sync and Sync to Exact Content Framerate on my machine.

**Comment by ehw** [comment]
Hello! 

I looked into the issue with flycast's frame pacing issues and I was able to conclude a few things, as least on my end. I apologize in advance if this might come across as redundant at this point...

I performed several tests using various core settings and was able to figure out a few things that do seem to affect the frame pacing. I performed these tests with the following:

Hardware:
1.) Intel 2500k overclocked to 4.2ghz.
2.) NVIDIA GeForce 680 GTX (slightly overclocked) with the 430.86 driver
3.) ASRock z68 Extreme 3 Gen 3 Motherboard.
4.) Dual monitor setup - Gsync enabled Acer XB271HU (DP, 120hz), Dell U2412M IPS monitor (DVI-D, 60hz)

Software:
1.) Windows 10 x64 1903 Education
2.) Retroarch (latest build bot as of 6/8/2019)
3.) Flycast (non WinCE core, latest as of 6/8/2019)
4.) 240p Test Suite cdi
5.) RivaTuner Statistics Server (for monitoring framerate and frametime)
6.) NVidia Shadowplay (for recording output from gpu)

I downloaded the latest Retroarch nightly as of 6/8/2019 and put it in a folder of it's own for testing. I made the following setting changes:
(Drivers)
Menu - ozone
(Video)
1.) Monitor Index - 1 (so that my Gsync monitor is default)
2.) Fullscreen - on
3.) Windowed fullscreen mode - off
4.) Vertical Refresh Rate - 120.000 Hz (reported by Display-Reported Refresh Rate)
5.) Vsync - On
(Frame throttle)
1.) Sync to Exact Content Framerate - On
(Onscreen Notification)
1.) Display Framerate - On (to report framerate as reported by Retroarch)

I downloaded the latest Flycast (non WinCE) core and modified it with the following settings initially:

> reicast_allow_service_buttons = "disabled"
> reicast_alpha_sorting = "per-strip (fast, least accurate)"
> reicast_analog_stick_deadzone = "15%"
> reicast_boot_to_bios = "disabled"
> reicast_broadcast = "NTSC"
> reicast_cable_type = "TV (Composite)"
> reicast_cpu_mode = "dynamic_recompiler"
> reicast_custom_textures = "disabled"
> reicast_digital_triggers = "disabled"
> reicast_div_matching = "disabled"
> reicast_dump_textures = "disabled"
> reicast_enable_dsp = "disabled"
> reicast_enable_purupuru = "disabled"
> reicast_enable_rtt = "disabled"
> reicast_enable_rttb = "disabled"
> reicast_frame_skipping = "disabled"
> reicast_framerate = "fullspeed"
> reicast_gdrom_fast_loading = "disabled"
> reicast_internal_resolution = "640x480"
> reicast_language = "Default"
> reicast_lightgun1_crosshair = "disabled"
> reicast_lightgun2_crosshair = "disabled"
> reicast_lightgun3_crosshair = "disabled"
> reicast_lightgun4_crosshair = "disabled"
> reicast_mipmapping = "disabled"
> reicast_oit_abuffer_size = "512MB"
> reicast_per_content_vmus = "disabled"
> reicast_region = "Default"
> reicast_render_to_texture_upscaling = "1x"
> reicast_screen_rotation = "horizontal"
> reicast_synchronous_rendering = "disabled"
> reicast_system = "auto"
> reicast_texupscale = "off"
> reicast_texupscale_max_filtered_texture_size = "256"
> reicast_threaded_rendering = "disabled"
> reicast_trigger_deadzone = "0%"
> reicast_vmu1_pixel_off_color = "DEFAULT_OFF 01"
> reicast_vmu1_pixel_on_color = "DEFAULT_ON 00"
> reicast_vmu1_screen_display = "disabled"
> reicast_vmu1_screen_opacity = "100%"
> reicast_vmu1_screen_position = "Upper Left"
> reicast_vmu1_screen_size_mult = "1x"
> reicast_vmu2_pixel_off_color = "DEFAULT_OFF 01"
> reicast_vmu2_pixel_on_color = "DEFAULT_ON 00"
> reicast_vmu2_screen_display = "disabled"
> reicast_vmu2_screen_opacity = "100%"
> reicast_vmu2_screen_position = "Upper Left"
> reicast_vmu2_screen_size_mult = "1x"
> reicast_vmu3_pixel_off_color = "DEFAULT_OFF 01"
> reicast_vmu3_pixel_on_color = "DEFAULT_ON 00"
> reicast_vmu3_screen_display = "disabled"
> reicast_vmu3_screen_opacity = "100%"
> reicast_vmu3_screen_position = "Upper Left"
> reicast_vmu3_screen_size_mult = "1x"
> reicast_vmu4_pixel_off_color = "DEFAULT_OFF 01"
> reicast_vmu4_pixel_on_color = "DEFAULT_ON 00"
> reicast_vmu4_screen_display = "disabled"
> reicast_vmu4_screen_opacity = "100%"
> reicast_vmu4_screen_position = "Upper Left"
> reicast_vmu4_screen_size_mult = "1x"
> reicast_volume_modifier_enable = "disabled"
> reicast_widescreen_hack = "disabled"

For this test, any time I made a setting change for the core, I closed and reopened the content itself. I recorded the frametime/framerate of the tests with Rivatuner refreshing the data on the graph every 500ms. I set the power management mode for retroarch.exe in the NVidia CP to Prefer Maximum Performance. I also set the Nvidia CP to enable an indicator when GSync has been engaged. I set Shadowplay to record at 1080p and at a framerate at 60fps.

Here are the results of the tests:
![image](https://user-images.githubusercontent.com/885618/59163872-46616a80-8ad4-11e9-8b17-b9b90b4a20eb.png)

You can view the recordings of each test here (note, if you watch it on certain media players on your PC, you'll get additional stuttering on top of the stuttering that was recorded. You'd have to find a way to view the video in something that allows you to watch it in exclusive fullscreen mode, but otherwise you should be able to still tell if stuttering is present in the video itself.:
[https://mega.nz/#F!tQpQDSzI!WW8rLxBGOVI6i4XMwStI0g](url)

In short:
1.) The Normal/Fullspeed setting doesn't seem to affect the frame pacing issue, at least as far as what I view as frame pacing. It probably has more of an effect in more demanding titles.

2.) To confirm what was stated before, it doesn't seem like there's any stuttering with either TV(RGB) or TV(Composite) cables when the video is set to 240p. If set to 480i however, the frame times go crazy and there's a lot of stuttering. I can understand this do to the nature of 480i being an interlaced video mode however...

3.) ...but using VGA produces dropped frames, even if the video is 480p - which doesn't make sense to me. If the video is progressive, like 240p, shouldn't the results be the same as 240p but at a higher resolution? Why are the frame timings so crazy here? In fact, the frame timings seems degenerative, getting gradually worse if you open the Retroarch menu or unfocusing the window and wait a few seconds. Watch "480p stutter graph anomaly.mp4" to see it happen. Any idea what causes this?

4.) Threaded and synchronous rendering produces massive stuttering that does NOT get reported by the RivaTuner time graph, but it's definitely there and you can see it in the recordings. While I know using synchronous rendering by itself doesn't do anything, just having threaded rendering enabled by itself is enough to produce these stutters.

5.) NTSC and PAL video modes are equally affected by the setting changes.

Again, I apologize if this strays away from this issue - perhaps this might warrant another one instead. But I do see some weird problems here, especially with the threaded renderer. 

I hope this was at least some help...

**Comment by ehw** [comment]
Wanted to give an update since a lot has changed since the last time I used flycast.

It seems that enabling Fullspeed and setting the emulator to TV(Composite) mode fixes the stuttering as it should. For some reason, I had to disable audio synchronization before in order to get it to display smoothly with Fullspeed, but now you no longer have to do that (at least for me)? Unfortunately, threaded and/or synchronous rendering still produces the frame pacing issues as seen with the Normal framerate setting, and there's no way to fix that at the moment. :(

I recorded two videos to show how bad the stuttering is between fullspeed and threaded/synchronous rendering. I recommend watching these in a video player that supports exclusive fullscreen.

Fullspeed (little to no dropped frames in video, none in real time):
https://mega.nz/#!oMQ01aoC!vP5pPUG9SYAEgQnA5CTy7Ur2PjPSVCas8Q4_lr3aN5E

Threaded/Synchronous video (lots of stuttering):
https://mega.nz/#!MVYE3SpA!4O1xBYkhx988PuZ_gVMFiVQHv1n68lY-wTBgxdbzf4Q

Hope this helps. These two videos were made with the vulkan renderer.

**Comment by Awakened0** [comment]
I was just testing the Vulkan renderer myself. It's excellent since I can now use per pixel alpha sorting on my Windows desktop without sound crackling. Even with max swap chain set to 2. Thanks @flyinghead !

With Vulkan most games seem smooth to me with the full speed framerate setting, but Bangai-O still has scrolling stutter. It's mostly eliminated in that game by setting it to normal framerate, as before. This is with threaded and synchronous off, audio sync and sync to exact framerate on with Gsync. I use a core options override to only use normal framerate (and 320x240) for that game.

**Comment by flyinghead** [comment]
Closing for now. Reopen as needed
